localstack/services/apigateway/helpers.py

```diff
@@ -108,6 +106,78 @@ def __init__(self):
         self.client_certificates = {}
 
 
+class Resolver:
+    def __init__(self, document, allow_recursive=True):
+        self.document = document
+        self.allow_recursive = allow_recursive
+        # cache which maps known refs to part of the document
+        self._cache = {}
+        self._refpaths = ["#"]
+
+    def _is_ref(self, item):
+        return isinstance(item, dict) and "$ref" in item
+
+    def _is_internal_ref(self, refpath):
+        return str(refpath).startswith("#/")
+
+    @property
+    def current_path(self):
+        return self._refpaths[-1]
+
+    @contextlib.contextmanager
+    def _pathctx(self, refpath):
+        if not self._is_internal_ref(refpath):
+            refpath = "/".join((self.current_path, str(refpath)))
+
+        self._refpaths.append(refpath)
+        yield
+        self._refpaths.pop()
+
+    def _resolve_refpath(self, refpath):
+        if refpath in self._refpaths and not self.allow_recursive:
+            raise Exception("recursion detected with allow_recursive=False")
+
+        if refpath in self._cache:
+            return self._cache[refpath]
+
+        with self._pathctx(refpath):
+            if self._is_internal_ref(self.current_path):
+                cur = self.document
+            else:
+                raise NotImplementedError("External references not yet supported.")
+
+            for step in self.current_path.split("/")[1:]:
+                cur = cur[step]
```

Comment:
nit: We could use `cur.get(step)` here to hedge against `KeyError`, and then introduce a custom exception (e.g., `UnresolvableRefException`) that callers can catch explicitly. 

The other option would be to handle these cases gracefully and return `None` from the method if a ref cannot be resolved. 

I'd chose one of the two (instead of bubbling up raw `KeyError`s), to make the interface a bit more clear/explicit.

Reply:
Good points for sure, I made the modifications to hedge against the lookup and that is sufficient because we want the unresolvable references to be `None`. If you notice in this method the lookup is only supported for local references (no external references) and if the `current_path` (e.g `#/definitions/NotFound`) is not in the document, with the new lookup, it will return `None` and we proceed with the rest of the document.